### ISSUE TITLE
Ignore battery optimisations dialog

### ISSUE THREAD

**Issue by JanuszChmiel** [zeljkoc/PandroidPackage#7]
Thanks to The author of this Pandroid bundle, every Pascal developer for Android can use The following code to enforce to display ignore battery optimisations settings dialog window.
   pm:= AOPowerManager(getSystemService(ACContext.POWER_SERVICE)) ;
 if (AOBuild.InnerVERSION.fSDK_INT >= AOBuild.InnerVERSION_CODES.M) then begin
 if (not pm.isIgnoringBatteryOptimizations(getPackageName())) then begin
 // Prompt the user to disable battery optimization
 intent := ACIntent.create;
 intent.setAction(APSettings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);

 startActivity(intent);

 end;
 end;
But this code recall complex form of The dialog. User must choose all apps and then user must choose his running app  name so Android Google developers have prepared more simple approach. System UI dialog which prompt user if he want to have ignore battery optimisations enabled for running app by pressing OK button.
But how to recall this more simple form of The dialog window in Free Pascal PPCJVM?
For An inspiration, Termux terminal emulator contain The following code inside termuxservice.java

        String packageName = getPackageName();
        if (!pm.isIgnoringBatteryOptimizations(packageName)) {
            Intent whitelist = new Intent();
            whitelist.setAction(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
            whitelist.setData(Uri.parse("package:" + packageName));
            whitelist.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            try {
                startActivity(whitelist);
            } catch (ActivityNotFoundException e) {
But how to transfer this code to Pascal? I Am strongly recommend to all of us to use AndroidR25.inc, because newer APi headers do not support overide when creating GUI by using R25 compatible style. You will get An errors related to overide commands.
Sure, everyone must add 
<uses-permission android:name="android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS"/>
permission in to androidmanifest.xml, this ie very important. I AM also compiling by using Android29 platform and fortunately, it is possible with no need to install other than preinstalled vversion of Java or Apache ANT.

**Comment by zeljkoc** [comment]
this is a translation from java:
var
  whitelist: ACIntent;
  pm: AOPowerManager;
begin
   pm:= AOPowerManager(getSystemService(ACContext.POWER_SERVICE)) ;

   if not pm.isIgnoringBatteryOptimizations(getPackageName) then begin
          whitelist := ACIntent.create;

whitelist.setAction(APSettings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
          whitelist.setData(ANUri.parse(JLString('package:' +
getPackageName)));
          whitelist.setFlags(ACIntent.FLAG_ACTIVITY_NEW_TASK);
      try
         startActivity(whitelist);
      except on e: ACActivityNotFoundException do
         AWToast.makeText(self, e.getMessage.toString
,AWToast.LENGTH_SHORT).show();
      end;
   end;

ned, 25. srp 2021. u 13:44 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh, you are wonderful and I will try to add this code to my radio player code. Thank you for yours help and time.

**Comment by zeljkoc** [comment]
did you look at this:

My suggestion is to change AWArrayAdapter and  and make a new guy like I
suggested  TArrayAdapter;

  TArrayAdapter = class(AWArrayAdapter)
  fContext: ACContext;
  fItems: JUList;
  public
    constructor create(para1: ACContext; para2: jint; para3: JUList);
overload;
    function getView(para1: jint; aView: AVView; aViewGroup: AVViewGroup):
AVView;  override;
  end;

constructor TArrayAdapter.create(para1: ACContext; para2: jint; para3:
JUList);
begin
  fContext := para1;
  fItems := para3;
  inherited create(fContext,  para2, para3);
end;

function TArrayAdapter.getView(para1: jint; aView: AVView; aViewGroup:
AVViewGroup): AVView;
var
 lObject: AWLinearLayout;
 tv: AWTextView;
 drawable: AGDGradientDrawable;
begin
   lObject:= AWLinearLayout.Create(fContext);
   lObject.setOrientation(AWLinearLayout.VERTICAL);

     drawable:= AGDGradientDrawable.create;
  with drawable do begin
    if ((para1 mod 2 = 0)) then setColor(AGColor.parseColor('#e0e0e0'))
                                        else
setColor(AGColor.parseColor('#FFFFFF'));
    setStroke(1,  AGColor.parseColor('#00897B'));   //rub prekidaca
    setCornerRadius($07);
  end;

     tv:= AWTextView.create(fContext);
     tv.setText(ATHtml.fromHtml(JLString(fitems.get(para1)).toString.trim));
     tv.setBackgroundDrawable(drawable );
     tv.Height := 50;
     tv.setGravity(AVGravity.CENTER);
   lObject.addView(tv);

  Result := lObject;
end;

ned, 25. srp 2021. u 15:13 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
But I do not know how to add text items with station names and how to detect a item number clicked. This is The reason Why I have used yours Gritview so I could know how to detect item clicked.
But thank you very very much for yours suggested code.

**Comment by zeljkoc** [comment]
the same code is just replace the new component


ned, 25. srp 2021. u 15:54 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh it is wonderful. I will try it.

**Comment by zeljkoc** [comment]
I put that you can write html code in names

items.add(JLString('<font color=''red''>Radio Wave<font>'));
 items.add(JLString('<b> Radiožurnál </b>'));

ned, 25. srp 2021. u 16:41 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh, thank you for yours explanation. Ignore battery optimisations simple dialog really works, no crash. Thank you and now, I will use yours proposed GUI. I had issue with my older table gritview code, so I will begin from scratch by using yours proposed code.
And thank you for yours kind support. . Fortunately for Me, Termuxservice have contained good Java code which recall The dialog and yours qualified brain have converted Java code to Pascal.

**Comment by zeljkoc** [comment]
add to type:

type

  { TArrayAdapter }

  TArrayAdapter = class(AWArrayAdapter)
  fContext: ACContext;
  fItems: JUList;
  public
    constructor create(para1: ACContext; para2: jint; para3: JUList);
overload;
    function getView(para1: jint; aView: AVView; aViewGroup: AVViewGroup):
AVView;  override;
  end;

//----------------------------------------------------------------


implementation

{ TArrayAdapter }

constructor TArrayAdapter.create(para1: ACContext; para2: jint; para3:
JUList);
begin
  fContext := para1;
  fItems := para3;
  inherited create(fContext,  para2, para3);
end;

function TArrayAdapter.getView(para1: jint; aView: AVView; aViewGroup:
AVViewGroup): AVView;
var
 lObject: AWLinearLayout;
 tv: AWTextView;
 drawable: AGDGradientDrawable;
begin
   lObject:= AWLinearLayout.Create(fContext);
   lObject.setOrientation(AWLinearLayout.VERTICAL);

     drawable:= AGDGradientDrawable.create;
  with drawable do begin
    if ((para1 mod 2 = 0)) then setColor(AGColor.parseColor('#e0e0e0'))
                                        else
setColor(AGColor.parseColor('#FFFFFF'));
    setStroke(1,  AGColor.parseColor('#00897B'));   //rub prekidaca
    setCornerRadius($07);
  end;

     tv:= AWTextView.create(fContext);
     tv.setText(ATHtml.fromHtml(JLString(fitems.get(para1)).toString.trim));
     tv.setBackgroundDrawable(drawable );
     tv.Height := 50;
     tv.setGravity(AVGravity.CENTER);
   lObject.addView(tv);

  Result := lObject;
end;

//--------------------------
change:
Adapter: TArrayAdapter;
and
Adapter := TArrayAdapter.create(Self, AR.innerLayout.simple_list_item_1,
items);

ned, 25. srp 2021. u 17:52 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Code works, but I need to add new battery optimisations code and I will send updated fully functioning code.
Thank you for yours detailed tutorial.

**Comment by JanuszChmiel** [comment]
Here is A complete code Thanks to yours programmers help. May be, that I should remove gritview related values for width, etc.
Did I place The codetogethetr correctly?

unit tabulka;

{$mode objfpc}{$H+}
{$modeswitch unicodestrings}
{$namespace pandroid.com.tabulka}
interface

uses androidr25, Rjava, AActivity, ZCBass;
type

{ TArrayAdapter }

TArrayAdapter = class(AWArrayAdapter)
fContext: ACContext;
fItems: JUList;
public
constructor create(para1: ACContext; para2: jint; para3: JUList);
overload;
function getView(para1: jint; aView: AVView; aViewGroup: AVViewGroup):
AVView; override;
end;



type

    { TPhonePlayer }

   TPhonePlayer = class(ATPhoneStateListener)
     FContext : ACContext;
     procedure onCallStateChanged(para1: jint; para2: JLString); overload; override;
      public
     property Context: ACContext read FContext write FContext;
   public

  end;

  { MainActivity }

  MainActivity = class(Activity, CUBBASS.InnerDOWNLOADPROC)

  public
    procedure onCreate(savedInstanceState: AOBundle); override;
    procedure onItemClick(aAdapter: AWAdapterView; aView: AVView; position: LongInt; id: Int64); override;
        procedure DOWNLOADPROC(para1: JNByteBuffer; para2: jint; para3: JLObject); overload;

  protected
     procedure Play(aUrl: JLString);
     procedure StopPlay;
     procedure StartPlay;

  public
    items: JUList;
        TelephonyManager : ATTelephonyManager;
    PhonePlayer: TPhonePlayer;
    lastURL: JLString;
    bass: CUBBASS;
    sample : jint;
  channel: jint;
  pm: AOPowerManager;
whitelist: ACIntent;

  public

  end;

implementation
{ TArrayAdapter }

constructor TArrayAdapter.create(para1: ACContext; para2: jint; para3:
JUList);
begin
fContext := para1;
fItems := para3;
inherited create(fContext, para2, para3);
end;

function TArrayAdapter.getView(para1: jint; aView: AVView; aViewGroup:
AVViewGroup): AVView;
var
lObject: AWLinearLayout;
tv: AWTextView;
drawable: AGDGradientDrawable;
begin
lObject:= AWLinearLayout.Create(fContext);
lObject.setOrientation(AWLinearLayout.VERTICAL);

drawable:= AGDGradientDrawable.create;
with drawable do begin
if ((para1 mod 2 = 0)) then setColor(AGColor.parseColor('#e0e0e0'))
else
setColor(AGColor.parseColor('#FFFFFF'));
setStroke(1, AGColor.parseColor('#00897B')); //rub prekidaca
setCornerRadius($07);
end;

tv:= AWTextView.create(fContext);
tv.setText(ATHtml.fromHtml(JLString(fitems.get(para1)).toString.trim));
tv.setBackgroundDrawable(drawable );
tv.Height := 50;
tv.setGravity(AVGravity.CENTER);
lObject.addView(tv);

Result := lObject;
end;


{ TPhonePlayer }

procedure TPhonePlayer.onCallStateChanged(para1: jint; para2: JLString);
begin
 // inherited onCallStateChanged(para1, para2);
 if (FContext is MainActivity) then
   case para1 of
    ATTelephonyManager.CALL_STATE_RINGING,
    ATTelephonyManager.CALL_STATE_OFFHOOK: (FContext as MainActivity).StopPlay;

    ATTelephonyManager.CALL_STATE_IDLE: (FContext as MainActivity).StartPlay;
  end;

end;


procedure MainActivity.Play(aUrl: JLString);
begin
    lastURL := aUrl;
  bass:= CUBBASS.create();

Bass.BASS_StreamFree(channel);
 channel := bass.BASS_StreamCreateURL(aUrl, 0, CUBBASS.BASS_DEVICE_DEFAULT, Self, self);
 Bass.BASS_ChannelPlay(channel, false);
  end;

procedure MainActivity.StartPlay;
begin
Play(LastUrl);
  end;

procedure MainActivity.StopPlay;
begin
lastURL :='';
Bass.BASS_StreamFree(channel);

  end;

procedure MainActivity.onCreate(savedInstanceState: AOBundle);
var
 layout: AWLinearLayout;
 params: AWLinearLayout.InnerLayoutParams;
 GridView: AWGridView;
 Adapter: TArrayAdapter;

begin
  inherited onCreate(savedInstanceState);
  layout:= AWLinearLayout.Create(self);
  params:= AWLinearLayout.InnerLayoutParams.Create(
              AWLinearLayout.InnerLayoutParams.FILL_PARENT ,
              AWLinearLayout.InnerLayoutParams.WRAP_CONTENT);
  layout.setOrientation(AWLinearLayout.HORIZONTAL);
layout.setGravity(AVGravity.CENTER);

   items:= JUArrayList.Create;
    items.add(JLString('Radiožurnál'));
                                                            items.add(JLString('Dvojka'));
    items.add(JLString('Vltava'));
    items.add(JLString('Plus'));
   items.add(JLString('D-dur'));
   items.add(JLString('Jazz'));
   items.add(JLString('Radio Wave'));
   items.add(JLString('Rádio Junior'));
   items.add(JLString('Rádio Junior písničky'));
   items.add(JLString('Radio Retro'));
   items.add(JLString('Rádio Junior Písničky'));
   items.add(JLString('Radio Praha'));
   items.add(JLString('Sport'));
   items.add(JLString('Brno'));
   items.add(JLString('České Budějovice'));
   items.add(JLString('Hradec Králové'));
   items.add(JLString('Olomouc'));
   items.add(JLString('Ostrava'));
   items.add(JLString('Pardubice'));
   items.add(JLString('Plzeň'));
   items.add(JLString('Karlovy Vary'));
   items.add(JLString('Regina'));
   items.add(JLString('Sever'));
   items.add(JLString('Liberec'));
   items.add(JLString('Region Střední Čechy'));
   items.add(JLString('Region Vysočina'));
  pm:= AOPowerManager(getSystemService(ACContext.POWER_SERVICE)) ;
if (AOBuild.InnerVERSION.fSDK_INT >= AOBuild.InnerVERSION_CODES.M) then begin
whitelist := ACIntent.create;

whitelist.setAction(APSettings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
whitelist.setData(ANUri.parse(JLString('package:' +
getPackageName)));
whitelist.setFlags(ACIntent.FLAG_ACTIVITY_NEW_TASK);
try
startActivity(whitelist);
except on e: ACActivityNotFoundException do
AWToast.makeText(self, e.getMessage.toString
,AWToast.LENGTH_SHORT).show();
end;
end;

   TelephonyManager :=  ATTelephonyManager(getSystemService(ACContext.TELEPHONY_SERVICE));
  PhonePlayer:= TPhonePlayer.create;
  PhonePlayer.Context := Self;
  TelephonyManager.listen(PhonePlayer, ATPhoneStateListener.LISTEN_CALL_STATE);
  bass.BASS_Init(-1, 44100, 0);
  GridView:= AWGridView.Create(Self);
  Adapter := TArrayAdapter.create(Self, AR.innerLayout.simple_list_item_1,
items);

GridView.setVerticalSpacing(24);
  GridView.setHorizontalSpacing(10);
  GridView.setColumnWidth(100);
  GridView.setNumColumns(AWGridView.AUTO_FIT);
  GridView.setOnItemClickListener(Self);
  GridView.setAdapter(Adapter);

  layout.addView(GridView);
  setContentView(layout);
end;

procedure MainActivity.onItemClick(aAdapter: AWAdapterView; aView: AVView;  position: LongInt; id: Int64);
begin

  case id of
        0: begin

          Play('http://icecast8.play.cz/cro1-128.mp3');
                           end;
        1: Play('http://icecast6.play.cz/cro2-128.mp3');
        2: Play('http://icecast5.play.cz/cro3-128.mp3');
        3: Play('http://icecast1.play.cz/croplus128.mp3');
4: Play('http://icecast5.play.cz/croddur-128.mp3');
5: Play('http://icecast1.play.cz/crojazz128.mp3');
6: Play('http://icecast5.play.cz/crowave-128.mp3');
        7: Play('http://icecast5.play.cz/crojuniormaxi128.mp3');
8: Play('http://icecast7.play.cz/crojuniormini128.mp3');
9: Play('http://ice3.abradio.cz/croretro128.mp3');
        10: Play('http://icecast2.play.cz/cro7-128.mp3');
        11: Play('http://icecast7.play.cz/crosport128.mp3');
        12: Play('http://icecast2.play.cz/crobrno128.mp3');
        13: Play('http://icecast2.play.cz/crocb128.mp3');
        14: Play('http://icecast2.play.cz/crohk128.mp3');
        15: Play('http://icecast2.play.cz/crool128.mp3');
        16: Play('http://icecast2.play.cz/croov128.mp3');
        17: Play('http://icecast2.play.cz/cropardubice128.mp3');
        18: Play('http://icecast2.play.cz/croplzen128.mp3');
        19: Play('http://icecast2.play.cz/crokv128.mp3');
        20: Play('http://icecast2.play.cz/croregina128.mp3');
        21: Play('http://icecast2.play.cz/crosever128.mp3');
        22: Play('http://icecast2.play.cz/croliberec128.mp3');
        23: Play('http://icecast2.play.cz/croregion128.mp3');
        24: Play('http://icecast2.play.cz/crovysocina128.mp3');
        end;
  end;

  procedure MainActivity.DOWNLOADPROC(para1: JNByteBuffer; para2: jint; para3: JLObject);
begin

end;
    end.

**Comment by JanuszChmiel** [comment]
Yes, when I have removed gritview related values, every station name is placed in one line, so I have one column and more than 20 lines, but lines are nicely placed, no so far between each others, space between lines is not so big so it is easily possible to switch between them with screen reader. Thank you for yours useful new GUI.

**Comment by zeljkoc** [comment]
Space:
GridView.setVerticalSpacing(3);
GridView.setHorizontalSpacing(3);

width of one row
GridView.setColumnWidth(150);

adjustment according to rotation
GridView.setNumColumns(AWGridView.AUTO_FIT);



ned, 25. srp 2021. u 20:55 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh, thank you very very much for yours tuned values. I will update The source according to it. Without values, Android format every station name on A separate line, so one column at every one line. I would be never able to calculate those values from my own head. And there is very probably no automated matematical formula available, it would be too complex to develop such formula, because it would had to involve The syze and font type and according to letters combination and according to screen in inches. Thank you again.

**Comment by JanuszChmiel** [comment]
Oh you are The onlyone man ewho have been able to tune column width and other values according to items. People in discussion forums around The world are so unpleasant to Me. Thank you very very much.

I have even found very interesting parameter related to gritview from Pandroid for Linux package Unfortunately, when I have tried to use it, Gritwiev contained empty data, no data, screen reader have only informed Me, that I AM using gritwiev but table have been empty. The interesting parameter is

GridView.setStretchMode(AR.Innerattr.COLUMNWIDTH);
I have tried it when I have called AndroidR25 headers.
You have even commented it by using // in The original gritwiev example on Pandroid for Linux. So it is something strange.

**Comment by JanuszChmiel** [comment]
I would like to publicly apologize to you that I have been too slow Yesterday. The reason was, that Windows defender have started to block external tools which have been used by Lazarus. The reason of doing that was, that I have run Lazarus as Administrator and in this case, Defender service and other security routines of Windows have been too careful. Fortunately, when I did not run Lazarus as administrator, those issue has stopped to occur. As A reason, I could not build .apk so fast. Yours programmers help have been very welcomed. And now I will do my best to be more independent. I will add some new GUi elements to The player. And I will be only pleasing you for tuned space values, because I can not do that myself, really.
Sure, also without tuned values, screen reader read everything. But resulting GUI look ugly. Sure, GUI would never look ugly even though I do not see at all, if I would only use yours azcforms. But because you do not have too much time and there are not too much volunteers, azcforms do not contain radio button support and checkbox support for example. So I have very big plea to you.
Because in reality, AZcforms contain reliable routines, which allow developer with no sight at all to create nice GUIS, values of objects positioning are not needed, could you consider to add more GUi elements? Sure, not all available on Android, it would be too much and I Am aware of that.

But if you could add radiobuttons support,and checkboxes support. The reason for my plea is, that I can not use Android APIS myself to do that, because I will always construct ugly GUIS. Because it is always necessary to specify object position values. Sure, I know about many samples, where I can add buttons by using Android APIS, for example in our first pandroid radioplayer based on Mediaplayer API.
But azcforms technique, which you are using for scrollable buttons for example or for buttons with logos is A good stratégy. But The question is, if it would be possible to add more GUI elements so users would never be forced to specify object position values.
So azcforms allow to add buttons without position values and GUi looks still nice.
How complex would be to add The similar routines for other GUI elements such as radiobutton, checkbox?
Thank You very much for yours time and for yours programmers debate.

**Comment by zeljkoc** [comment]
analyze a little the code I sent:

 lObject: AWLinearLayout;   - is one button
I put:  tv: TektViev; dravable: GradientDravable;
you can put an image or whatever you want in that object:

//-------------------------

function TArrayAdapter.getView(para1: jint; aView: AVView; aViewGroup:
AVViewGroup): AVView;
var
 lObject: AWLinearLayout;
 tv: AWTextView;
 drawable: AGDGradientDrawable;
begin
   lObject:= AWLinearLayout.Create(fContext);
   lObject.setOrientation(AWLinearLayout.VERTICAL);

     drawable:= AGDGradientDrawable.create;
  with drawable do begin
    if ((para1 mod 2 = 0)) then setColor(AGColor.parseColor('#e0e0e0'))
                                        else
setColor(AGColor.parseColor('#FFFFFF'));
    setStroke(1,  AGColor.parseColor('#00897B'));   //rub prekidaca
    setCornerRadius($07);
  end;

     tv:= AWTextView.create(fContext);
     tv.setText(ATHtml.fromHtml(JLString(fitems.get(para1)).toString.trim));
     tv.setBackgroundDrawable(drawable );
     tv.Height := 50;
     tv.setGravity(AVGravity.CENTER);
   lObject.addView(tv);

  Result := lObject;

pon, 26. srp 2021. u 10:53 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh, so it is excellent. Even text strings? Yours Android Api knowledge is really outstanding. Thank you for yours tip.

**Comment by JanuszChmiel** [comment]
And I have also decided to deeply study yours units and examples. I have also found out, that azcform can allow me to add text item. So I have neverendless possibilityes, which are only limited by my brain.
You have made big work, when you have prepared AndroidR25 Api so professionally, because overide can be used. So I can only tell you. Very well done and thank you.

**Comment by JanuszChmiel** [comment]
But I have found out very sad fact. Czech Unicode characters wirth diacritics can not be drawn if Jlstring data type is being used. So as A paradox. When I have used azcforms, all Czech characters have been displaied correctly on The display. When I have used buttons. But sure, when I used layout.addText ();
Czech characters have not been displaied correctly, because if Jlstring data type is being used, it do not support Czech Unicode characters.
I will look at algoritmus which have drawn buttons or may be, that you will give Me a piece of advice.

**Comment by zeljkoc** [comment]
it should do that, make an edit field and write it and then switch it to
azcforms

sri, 28. srp 2021. u 13:27 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Yes, I will do my best to extend yours Azcforms to several GUI elements such as checkbox, radiobutton, listview. If I will have some success, I will send you updated azcforms.pas unit to you.
I have started to manuály extract all object GUi widget related object class names to A separate new .txt file. Unfortunately I must confirm, that if AN widget text is drawn by using Jlstring data type, Czech encoding is always lost no matter which encoding I use. UTF8 do not work, ESE-ansi also not. Fortunately, azcforms creates Czech characters so I will do my best to extend it and if I will have some success, many other developers can benefit from this azcforms extension.
Do not be afraid, I will only write you my successfull result. You would only had to solve some hydden layout related issue.

**Comment by zeljkoc** [comment]
Look at the example, it's old, but it has syntax:
https://github.com/zeljkoc/pandroid/tree/master/example/MultiLanguage

language support is defined in
https://github.com/zeljkoc/pandroid/tree/master/example/MultiLanguage/android/res

the language depends on the suffix:
values-hr (hrvatski)
values-sl (slovenski)
values-zh (kineski)

i think your language is with -ch

in the citas program:
R.string_.rec

defined:
string.xml

<string name="Rec">Pozdravljen, svet! res/values-sl </string>


sri, 28. srp 2021. u 14:36 JanuszChmiel ***@***.***> napisao
je:

**Comment by JanuszChmiel** [comment]
Oh it would be so nice so I would be able to use also all methods which depend on Jlstring. Thank you very much for yours patience. Now I Am collecting all widget class names and I have decided that I will do my best to create Excel table which will containall widgets names and their functions and procedures so I will be able to find those information faster than by reading long androidr25.inc or other Api headers as A plain text. It will be a adventure work. If I will success, I will send a .xls table to you.

**Comment by JanuszChmiel** [comment]
Yes, this is ideal multilocalisation stratégy used also with Java language or Eclipse or with Kotlin. But Czech characters are correctly displaied when I Am using azcforms, but addbutton procedure also uses Jlstring function to convert strings to JLstrings and there is no problém. In this case, there is some encoding mismath produced by my use of Notepad and Lazarus at The same time.
I have tried to remove Czech characters and I have retyped them again from Lazarus source editor. It is really intresting phenomena. Because Jlstring must be used very often while calling Api functions for Android.
I will try to find The cause, it is interesting problém.
